### ISSUE TITLE
Logout route

### ISSUE THREAD

**Issue by blorange2** [codegreencreative/laravel-samlidp#61]
I have a query regarding the logout route, from my service provider I hit logout and this goes to `https://myapp.com/logout` but when it hits the logout route I get a 405 error as the default logout expects a POST request.

Does this mean I've set something up wrongly as this would imply that your listener is not firing.

```

  
<?php

namespace CodeGreenCreative\SamlIdp\Listeners;

use Illuminate\Auth\Events\Logout;

class SamlLogout
{
    /**
     * Upon logout, initiate SAML SLO process for each Service Provider
     * Simply redirect to the saml/logout route to handle SLO
     *
     * @param  Logout $event
     * @return void
     */
    public function handle(Logout $event)
    {
        // Make sure we are not in the process of SLO when handling the redirect
        if (in_array($event->guard, config('samlidp.guards')) && null === session('saml.slo')) {
            abort(redirect('saml/logout'), 200);
        }
    }
}

```

![image](https://user-images.githubusercontent.com/13929733/117334274-b7ad2200-ae91-11eb-9ae5-69b6b48dca78.png)

To me, it looks like it should just redirect to `saml/logout`, however I get a 405 error.

**Comment by upwebdesign** [comment]
@blorange2 Did you try to use `https://idp.com/saml/logout` route instead?

**Comment by blorange2** [comment]
@upwebdesign no, but only because I knew if I hit the logout route it would abort and do `saml/logout` anyway. I need to dig further into the package as I think its really cool but if I were to set the SP up just to go straight to `saml/logout` do you think that would work? Additionally, in my case is it because I am not sending the initial POST request to my default logout route?

**Comment by upwebdesign** [comment]
@blorange2 All my service providers link to `saml/logout`. In your IdP, you can post to your logout route and it should start the SLO as well.

**Comment by blorange2** [comment]
@upwebdesign ah, so as long as something triggers a post request, it'll fire? How do differentiate this for when users may just be logging out of the IDP? I swear I don't have many more questions.

**Comment by blorange2** [comment]
@upwebdesign maybe I'm misunderstanding how the logout process actually works. So, can I just doubly confirm my understanding is correct.

- An SP hits my logout route
- This triggers your SLO listener
- Your listener redirects to `saml/logout` which links to the index function in the `SamlLogoutController`

**Comment by OPVL** [comment]
```php
    public function saml2logout(Request $request)
    {
        $request->session()->pull('_rootinc_saml2_id');

        return redirect()->away($this->getLogoutUrl());
    }

    public function logoutcallback(Request $request)
    {
        $auth = $this->getAuth();

        // destroy the local session by firing the Logout event
        $keep_local_session = false;
        $retrieveParametersFromServer = false; //originally configurable in saml2_settings

        $auth->processSLO($keep_local_session, null, $retrieveParametersFromServer);
        $errors = $auth->getErrors();

        if (!empty($errors))
        {
            $this->logoutfail($request, $errors);
        }

        return $this->logoutsuccess($request);
    }
```
I'm not sure if this fully answers your question, I'm setting up an IDP using this package to integrate with [rootinc/laravel-saml2-middleware](https://github.com/OPVL/laravel-saml2-middleware) in my SP. Above is the code from that processes the logout in my SP.

### SP .env
```env
SAML2_IDP_SSO=http://idp.com/login
SAML2_IDP_SLO=http://idp.com/saml/logout

SAML2_SP_SSO=http://sp.com/login/saml2
SAML2_SP_SLO=http://sp.com/logout/saml2
```

### SP Routes
```php
Route::get('/saml2/metadata', '\RootInc\LaravelSaml2Middleware\Saml2@saml2metadata');
Route::get('/login/saml2', '\RootInc\LaravelSaml2Middleware\Saml2@saml2');
Route::post('/login/saml2callback', '\RootInc\LaravelSaml2Middleware\Saml2@saml2callback');
Route::get('/logout/saml2', '\RootInc\LaravelSaml2Middleware\Saml2@saml2logout');
Route::post('/logout/saml2callback', '\RootInc\LaravelSaml2Middleware\Saml2@logoutcallback');
```

### When a user wants to sign out
1.  they would be directed to sp.com/logout/saml2
2. SP then redirects to appropriate defined SLO route (away to IDP)
3. IDP handles incoming logout
4. IDP redirects user back to SP logout route where SLO logout is processed 

alternatively the IDP can 'touch' the logout routes for the other linked SPs and that would process the logout events as if your user is returning to the site on a normal login request.

i.e when user signs out of SP1 through IDP, IDP redirects user back to origin SP logout (if defined or specified in request), IDP posts to logout callback of SP2, SP2... and the user is de-authenticated across all SPs & IDP.

**Comment by upwebdesign** [comment]
@OPVL thanks for providing some insight on how you are implementing this package!

**Comment by blorange2** [comment]
@OPVL what is your SP doesn't have a logout URL? In my case I have to provide them a URL they can use to trigger the logout, so I gave them `saml/logout` 

When inspecting the flow this is what happens.

![image](https://user-images.githubusercontent.com/13929733/118293103-3711a580-b4d1-11eb-905b-c5168401e9c4.png)

So, would I in fact be better giving them a different URL?

This is my current saml config.

```

<?php

return [
    /*
    |--------------------------------------------------------------------------
    | SAML idP configuration file
    |--------------------------------------------------------------------------
    |
    | Use this file to configure the service providers you want to use.
    |
     */
    // Outputs data to your laravel.log file for debugging
    'debug' => true,
    // Define the email address field name in the users table
    'email_field' => 'email',
    // The URI to your login page
    'login_uri' => 'wub',
    // Log out of the IdP after SLO
    'logout_after_slo' => env('LOGOUT_AFTER_SLO', false),
    // The URI to the saml metadata file, this describes your idP
    'issuer_uri' => 'saml/metadata',
    // Name of the certificate PEM file
    'certname' => 'cert.pem',
    // Name of the certificate key PEM file
    'keyname' => 'key.pem',
    // list of all service providers
    'sp' => [
        'aHR0cHM6Ly9zZXJ2aWNlcy11ay5zdW5nYXJkZHguY29tL1NpbmdsZVNpZ25Pbi9TZXJ2aWNlUHJvdmlkZXI=' => [
            'destination' => 'https://services-uk.sungarddx.com/SingleSignOn/ServiceProvider',
            'logout' => 'saml/logout',
        ],
    ],

    // If you need to redirect after SLO depending on SLO initiator
    // key is beginning of HTTP_REFERER value from SERVER, value is redirect path
    'sp_slo_redirects' => [
        // 'https://example.com' => 'https://example.com',
        'https://services-uk.sungarddx.com/' => 'https://investment.newable.xyz/dashboard',
    ],
];

```

**Comment by blorange2** [comment]
@upwebdesign so I think the reason I'm having this issue is because my logout route configured in my SP is `saml/logout` so if you don't set a redirect you get stuck in a loop until it eventually redirects to the login route.

Also, my redirecting to `https://investment.newable.xyz/dashboard` seems to start a loop. I think this is because it redirects there but can't actually access it, so it just gets stuck in a loop of sorts.

When returning to the IDP is the session still intact?

**Comment by blorange2** [comment]
@upwebdesign in relation to the rest of these things (and apologies for asking so much), I was reading this article: https://www.identityserver.com/articles/the-challenge-of-building-saml-single-logout#:~:text=A%20SAML%20logout%20request%20follows,logging%20out%20the%20correct%20user. 

In your package you perform SSO via a logout URL with the IDP being the issuer. 

In my case, if the SP doesn't have a dedicated SLO URL, would i be better leaving it blank?

To be honest I think my SP could just be misunderstanding my questions, but I have asked them if they have an SLO endpoint.

Today, I read through the SLO job in this package and it would seem that the redirect loop is because I'm pointing to the IDP logout route.

P.S when I finally figure this out you'll have to set up a donation link so I can get you a coffee or something.

**Comment by upwebdesign** [comment]
@blorange2 if your SP has a logout route already, maybe you need to provide them with the SLO URL to your IdP? I have found that most SP's (that I have no control over) have their own logout route that is not changeable, meaning they are not set up for SLO. Facebook, for example, if you log out, you are logging out of Facebook and not your IdP. LinkedIn, however, has an SLO process that you set up in their administration section for SSO. Again, if your SP does not specifically have an SLO route, it likely wont work the way you want.

> In my case, if the SP doesn't have a dedicated SLO URL, would i be better leaving it blank?

If they don't have an SLO URL, then they do not support SLO, so yes, I would leave it blank. If you don't, you will never get a proper response back from their service.

In either case, the SP needs to hit your SLO logout route for it to work.

> P.S when I finally figure this out you'll have to set up a donation link so I can get you a coffee or something.

Appreciate that though! Hope this helps.

**Comment by blorange2** [comment]
@upwebdesign okay I think this is my final round of questioning

**As my SP has no dedicated URL would I leave this part empty?**

```

    'sp' => [
        'aHR0cHM6Ly9zZXJ2aWNlcy11ay5zdW5nYXJkZHguY29tL1NpbmdsZVNpZ25Pbi9TZXJ2aWNlUHJvdmlkZXI=' => [
            'destination' => 'https://services-uk.sungarddx.com/SingleSignOn/ServiceProvider',
            'logout' => '',
        ],
    ],

```

The `logout` key.

**If there is no logout key, we never have to trigger the SLO job?**

Looking at the source code of the job here: https://github.com/codegreencreative/laravel-samlidp/blob/master/src/Jobs/SamlSlo.php - if there is no `sp['logout'] the job will basically do nothing.

Further to this, in https://github.com/codegreencreative/laravel-samlidp/blob/master/src/Http/Controllers/LogoutController.php if I hit this controller it'll just redirect to my login page right?

**Is there a reason redirecting to login would break?**

So, as above I removed the logout key. Then I get this in my tracer.

![image](https://user-images.githubusercontent.com/13929733/118644949-3a10dc80-b7d6-11eb-91b9-e0cd8189cdf2.png)

It hits my login URL attempts to log back in and then redirects back to the login, sending the following SAML.

![image](https://user-images.githubusercontent.com/13929733/118645233-91af4800-b7d6-11eb-8ff2-20a41b79c703.png)

But then this happens.

![image](https://user-images.githubusercontent.com/13929733/118645334-b4d9f780-b7d6-11eb-8cc0-fca7d688f13d.png)

**Comment by upwebdesign** [comment]
@blorange2 Do you have a password manager attempting to log you in after logout?

**Comment by upwebdesign** [comment]
@blorange2 you can comment out the logout key altogether.

**Comment by blorange2** [comment]
@upwebdesign I don't believe so? Could it be that I've got event mappings for both Login and Authenticated?

![image](https://user-images.githubusercontent.com/13929733/118666566-320f6780-b7eb-11eb-8344-725e2ed9e711.png)

I could disable LastPass and try in an incognito window?

It does seem like there's a double login attempt.

**Comment by upwebdesign** [comment]
@blorange2 You definitely cannot post to the login route without a CSRF token. I just viewed the process for my IdP and SP's and here is what I found.

When attempting to log out at the IdP:

- POST https://idp.test/logout (doesn't end the session yet)
- GET https://idp.test/saml/logout (starts looping through the SP's)

- GET logout route for 1st SP including a `SAML_Request`
- GET https://idp.test/saml/logout (has `SAM_Response`)
- Repeat previous 2 steps for any other SP's

- GET https://idp.test/login (once `saml/logout` realizes there are no other SP's to log out of get the login page)

There should be nothing to modify on your end accept the `@samlidp` directive in your login form.

If you removed the `logout` keys from your config file, there would be no logout routes to redirect, therefore all you would be doing is logging out of your application and not any SP's. This is assuming you have only one SP configured.

**Comment by blorange2** [comment]
@upwebdesign 

Okay, here are the key differences:

1. When clicking logout at my SP it goes to: https://services-uk.sungarddx.com/LogOn/LogOff?redirectUrl=https%3A%2F%2Fservices-uk.sungarddx.com%2FDocument%2F93203 
2. This then does a GET request to `saml/logout` 
3. Due to the `redirectUrl` parameter in my SP's logout, it redirects back to where we were originally (https://services-uk.sungarddx.com/Document/93203)
4. Then, because we're no longer authenticated it tries to do a POST request to my login route

If there's a CSRF I get the token mismatch exception.
If I exclude `/login` from CSRF protection I get an error when it tries to use gzinflate

I would assume this is because it doesn't like the format of the data being sent?

So I suppose my question would be for my SP? As in, what are you trying to POST to my login route?

Just for clarity there is no way I'm leaving `/login` with the CSRF token...

Further to this I think where my SP is going wrong is as follows:

1. Sending a GET request to `saml/logout`
2. Including a `redirectUrl` for some unknown reason

In a nutshell according to all of our discussion, the final stop should be a GET request to `/login`.

When I'm helping third parties to configure SP's, did I go wrong in not specifying the types of requests? 

I ask as the setup they have seems extremely different to everything I've seen at this point.

As you can probably tell, all I've been doing for basically a week is reading up on SAML.

**Comment by upwebdesign** [comment]
@blorange2 its fine to test your login route without CSRF protection, but never leave it that way (I see you agree). Your SP should never send a POST to your login route. There is a lot to know about SAML and not just from an IdP view, but also an SP. I have found myself guiding MOST SP's and educating them on SAML most of the time. There are few instances where they use SAML properly out the gate. It sounds like your SP is on the fritz and needs guidance.

**Comment by blorange2** [comment]
@upwebdesign that makes sense. At this point I think its finally time to close this issue, as its not in fact an issue with either your package, or any parts of its implementation. I don't know whether it would be worth adding a section to the README about caveats with Service Providers?

This is mainly due to the fact that all of my issues have actually stemmed from the SP setup. I'd be happy to add this section if you think its appropriate?

**Comment by upwebdesign** [comment]
@blorange2 if you want to take a stab at the readme update, have it it!
